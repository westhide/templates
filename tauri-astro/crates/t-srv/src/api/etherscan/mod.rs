pub mod block;
pub mod client;
pub mod error;
pub mod transaction;
