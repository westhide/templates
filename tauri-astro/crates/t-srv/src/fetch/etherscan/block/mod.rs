pub mod get_number;
