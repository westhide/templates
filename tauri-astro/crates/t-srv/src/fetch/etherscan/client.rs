use std::{
    ops::{Deref, DerefMut},
    time::Duration,
};

use alloy_chains::Chain;
use foundry_block_explorers::Client;
use rand::random_range;

// https://api.etherscan.io/v2/api?chainid=56&module=contract&action=getabi&address=0x0e7779e698052f8fe56c415c3818fcf89de9ac6d&apikey=K4SDMH5SKWHRNK6G3PIXH1UTCWS1RND9DN

const API_KEY: &[&str] = &[
    // default
    "K4SDMH5SKWHRNK6G3PIXH1UTCWS1RND9DN",
    // txlist
    "1BD7WRF9C2TPVB38TCVY9U1R6F2BVFRM8D",
    "6NTVDK2AAP433UNVGAGD3X4YR25GRXW1I4",
    "M2193Y7ZXYRSJYGZEDC8N2MA54JJUB87ZA",
    "DHIIQTBNFU7JTDZNFHSJJKRIHI87Y41ZNG",
    "PAM5T2NCRJESWY9TV96AUF29YBPNE895IE",
    "AYB7PI37YKAKU84C9B4AIKY2PWRVUNSZG3",
    "QQ4TC8B3Z3F91MJWR95N1YTPZVJMZWQQ3C",
    "6B7PWZP4Z4SDN61B58VW1GHBVATHRFPH4H",
    "U1UMTGEDSQSHCH1M2JYPCVWKT1VT5XA7TK",
    "3ZCQ31SJJIJHTWM98QZW5KK18KKC4E7UZM",
    "82YCUG2S2J1FCM1QERU8PIEKA9CGCH9VBE",
    "X3YGWTZ6IUVUTDEZ13IDZBK98G7DDIV76H",
    // txlistinternal
    "WU6BJG8GEZFRQXHHS4NWAGUXFQ1NV5VQKY",
    "1S74F22M1R4VZNYAJM6FAE46UDVFTCQIUB",
    "PURC6YBDH9R845NP3YCHUT4CIT1E8NB3XF",
    "TZCZNHGWBHEZ13XYC5JNCT639UBH7PDRKG",
    "VJ96HWYYURF6VYHX1EI53U6PNZQNTHRISJ",
    "4DGG9Z1C1J2EGRQCSDUCENM7SKDUGDAF6N",
    "TNVGQD5H1B46VKGYVXU7823P3V3GNA2N37",
    "FQIUSS616PPA3BXIGV5VM3YYKCKSRYXVRS",
    "3I4NKA8YTPQCDV2NU9D76EMUU67CNNVQE7",
    "FS2251G4TVG54DGF5HZ4A4X33E9IITAW9V",
    "6AUFM7CJIS1RPCRWVUIRP59XGJJHPTUXMV",
    "KATCVPX9JZJ6A9RSV7D5SHC5AZEYC4ZH3R",
    // tokentx
    "Y5X8M19FJT8F1Q74A64C11WHGHCFXBY827",
    "PQ7WTNE1G94KAQHPG7TBNUQFQCK451IGCV",
    "Z175JK283IND6VV785RSF33HEES6E9PG3Y",
    "AUD8N76NDUMH4GG2J46CB8RVU9FZ8465E6",
    "UZ2BWTSK9D385PYINEN36D77WV96DX9639",
    "H859Z4W5FFG8MQRE6MHV6QRYCPIUUQ61GP",
    "C99ISPCTBGY4VKAJQ5ZHAQ8IPFWBYVHRIG",
    "A94RSZYSCW9QCS79BNZH1S8DA9KM2ENAEJ",
    "TREFDHHPJY7UFPSTZXSI5PX9TE3VB5S1X7",
    "T4S323ETPFFTT4WGKM3NNV4Z7P72I1YX1H",
    "RW3T58UUIIE3TE66ZRXH4EATKN2GFZN7VB",
    "7TN7C9WME1783WZZYAJSJFRHCR9ZBVUD45",
];

pub struct EtherscanClient {
    client: Client,
}

impl EtherscanClient {
    pub fn init() -> Client {
        let idx = random_range(..API_KEY.len());
        let key = API_KEY[idx];
        let bsc = Chain::bsc_mainnet();
        let timeout = Duration::from_secs(30);

        let http = reqwest::ClientBuilder::new()
            .connect_timeout(timeout)
            .timeout(timeout)
            .build()
            .expect("Http Client");

        Client::builder()
            .with_api_key(key)
            .chain(bsc)
            .expect("BSC Chain")
            .with_client(http)
            .build()
            .expect("Etherscan Client")
    }

    pub fn new() -> Self {
        Self { client: Self::init() }
    }
}

impl Deref for EtherscanClient {
    type Target = Client;

    fn deref(&self) -> &Self::Target {
        &self.client
    }
}

impl DerefMut for EtherscanClient {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.client
    }
}
