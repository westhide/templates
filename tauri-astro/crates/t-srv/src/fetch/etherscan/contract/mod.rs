pub mod get_abi;
