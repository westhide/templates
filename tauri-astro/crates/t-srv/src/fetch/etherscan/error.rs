use std::{io::Error as StdIoError, sync::PoisonError};

use alloy::{hex::FromHexError, primitives::utils::UnitsError};
use foundry_block_explorers::errors::EtherscanError;
use serde_json::Error as SerdeJsonError;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    StdIoError(#[from] StdIoError),

    #[error(transparent)]
    EtherscanError(#[from] EtherscanError),

    #[error(transparent)]
    FromHexError(#[from] FromHexError),

    #[error(transparent)]
    UnitsError(#[from] UnitsError),

    #[error(transparent)]
    SerdeJsonError(#[from] SerdeJsonError),

    #[error("{0}")]
    Generic(String),
}

impl<T> From<PoisonError<T>> for Error {
    fn from(err: PoisonError<T>) -> Self {
        Self::Generic(err.to_string())
    }
}

impl From<Error> for t_lib::error::Error {
    fn from(err: Error) -> Self {
        Self::Generic(err.to_string())
    }
}

macro_rules! err {
    ($($arg:tt)*) => {
        Err($crate::fetch::etherscan::error::Error::Generic(format!($($arg)*)))
    }
}

#[allow(unused_imports)]
pub(crate) use err;

pub type Result<T, E = Error> = std::result::Result<T, E>;
