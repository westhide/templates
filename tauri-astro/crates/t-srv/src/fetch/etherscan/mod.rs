use crate::fetch::{Fetch, etherscan::client::Etherscan};

pub mod block;
pub mod client;
pub mod contract;
pub mod model;
pub mod transaction;

pub trait EtherscanFetch {
    type Ret;

    type Err;

    fn fetch(self) -> impl Future<Output = Result<Self::Ret, Self::Err>>;
}

impl<T> EtherscanFetch for T
where
    Etherscan: Fetch<Self>,
{
    type Err = <Etherscan as Fetch<Self>>::Err;
    type Ret = <Etherscan as Fetch<Self>>::Ret;

    async fn fetch(self) -> Result<Self::Ret, Self::Err> {
        Etherscan::new().fetch(self).await
    }
}
