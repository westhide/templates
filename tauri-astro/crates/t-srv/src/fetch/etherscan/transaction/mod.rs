pub mod get_erc20_token_transfer_events;
pub mod get_internal_transactions;
pub mod get_transactions;
pub mod sort;
