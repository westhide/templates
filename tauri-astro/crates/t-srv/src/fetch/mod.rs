pub mod etherscan;

// FIXME: use #![feature(inherent_associated_types)]
pub trait Param {
    type Ret;

    type Err;
}

pub trait Fetch<T> {
    type Ret;

    type Err;

    fn fetch(&mut self, params: T) -> impl Future<Output = Result<Self::Ret, Self::Err>>;
}
