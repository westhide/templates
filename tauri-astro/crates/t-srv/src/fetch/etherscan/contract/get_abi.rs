use alloy::{json_abi::JsonAbi, primitives::Address};
use serde::{Deserialize, Serialize};
use t_lib::log::{Level, instrument};

use crate::fetch::{
    Fetch, Param,
    etherscan::{client::EtherscanClient, error::Error},
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Params {
    pub contract: Address,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Data {
    pub block_number: usize,
}

impl Param for Params {
    type Err = Error;
    type Ret = JsonAbi;
}

impl Fetch<Params> for EtherscanClient {
    type Err = <Params as Param>::Err;
    type Ret = <Params as Param>::Ret;

    #[instrument(level = Level::TRACE, skip_all, err, fields(?params))]
    async fn fetch(&mut self, params: Params) -> Result<Self::Ret, Self::Err> {
        let Params { contract } = params;
        let abi = self.contract_abi(contract).await?;
        Ok(abi)
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use insta::assert_debug_snapshot;
    use nill::{Nil, nil};

    use super::*;
    use crate::fetch::etherscan::{EtherscanFetch, error::Result};

    const ULTI_TOKEN: &str = "0x0e7779e698052f8fe56c415c3818fcf89de9ac6d";

    fn save_abi(path: &str, abi: &JsonAbi) -> Result<Nil> {
        let json = serde_json::to_string(&abi)?;
        fs::write(path, json)?;
        Ok(nil)
    }

    #[tokio::test]
    async fn test_get_abi_ulti() -> Result<Nil> {
        let param = Params { contract: ULTI_TOKEN.parse()? };
        let abi = param.fetch().await?;

        assert_debug_snapshot!(&abi);
        save_abi("abi/ulti.json", &abi)?;
        Ok(nil)
    }
}
