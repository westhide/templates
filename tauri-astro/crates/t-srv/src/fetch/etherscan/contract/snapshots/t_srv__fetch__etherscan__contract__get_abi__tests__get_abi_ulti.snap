---
source: crates/t-srv/src/fetch/etherscan/contract/get_abi.rs
expression: "&abi"
---
JsonAbi {
    constructor: Some(
        Constructor {
            inputs: [
                Param {
                    ty: "string",
                    name: "name_",
                    components: [],
                    internal_type: Some(
                        Other {
                            contract: None,
                            ty: "string",
                        },
                    ),
                },
                Param {
                    ty: "string",
                    name: "symbol_",
                    components: [],
                    internal_type: Some(
                        Other {
                            contract: None,
                            ty: "string",
                        },
                    ),
                },
                Param {
                    ty: "uint256",
                    name: "cap_",
                    components: [],
                    internal_type: Some(
                        Other {
                            contract: None,
                            ty: "uint256",
                        },
                    ),
                },
            ],
            state_mutability: NonPayable,
        },
    ),
    fallback: None,
    receive: None,
    functions: {
        "allowance": [
            Function {
                name: "allowance",
                inputs: [
                    Param {
                        ty: "address",
                        name: "owner",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "address",
                        name: "spender",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
                outputs: [
                    Param {
                        ty: "uint256",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "approve": [
            Function {
                name: "approve",
                inputs: [
                    Param {
                        ty: "address",
                        name: "spender",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "value",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                outputs: [
                    Param {
                        ty: "bool",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "bool",
                            },
                        ),
                    },
                ],
                state_mutability: NonPayable,
            },
        ],
        "balanceOf": [
            Function {
                name: "balanceOf",
                inputs: [
                    Param {
                        ty: "address",
                        name: "account",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
                outputs: [
                    Param {
                        ty: "uint256",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "burn": [
            Function {
                name: "burn",
                inputs: [
                    Param {
                        ty: "uint256",
                        name: "value",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                outputs: [],
                state_mutability: NonPayable,
            },
        ],
        "burnFrom": [
            Function {
                name: "burnFrom",
                inputs: [
                    Param {
                        ty: "address",
                        name: "account",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "value",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                outputs: [],
                state_mutability: NonPayable,
            },
        ],
        "cap": [
            Function {
                name: "cap",
                inputs: [],
                outputs: [
                    Param {
                        ty: "uint256",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "decimals": [
            Function {
                name: "decimals",
                inputs: [],
                outputs: [
                    Param {
                        ty: "uint8",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint8",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "mint": [
            Function {
                name: "mint",
                inputs: [
                    Param {
                        ty: "address",
                        name: "to",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "amount",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                outputs: [],
                state_mutability: NonPayable,
            },
        ],
        "name": [
            Function {
                name: "name",
                inputs: [],
                outputs: [
                    Param {
                        ty: "string",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "string",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "owner": [
            Function {
                name: "owner",
                inputs: [],
                outputs: [
                    Param {
                        ty: "address",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "renounceOwnership": [
            Function {
                name: "renounceOwnership",
                inputs: [],
                outputs: [],
                state_mutability: NonPayable,
            },
        ],
        "symbol": [
            Function {
                name: "symbol",
                inputs: [],
                outputs: [
                    Param {
                        ty: "string",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "string",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "totalSupply": [
            Function {
                name: "totalSupply",
                inputs: [],
                outputs: [
                    Param {
                        ty: "uint256",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                state_mutability: View,
            },
        ],
        "transfer": [
            Function {
                name: "transfer",
                inputs: [
                    Param {
                        ty: "address",
                        name: "to",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "value",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                outputs: [
                    Param {
                        ty: "bool",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "bool",
                            },
                        ),
                    },
                ],
                state_mutability: NonPayable,
            },
        ],
        "transferFrom": [
            Function {
                name: "transferFrom",
                inputs: [
                    Param {
                        ty: "address",
                        name: "from",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "address",
                        name: "to",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "value",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                outputs: [
                    Param {
                        ty: "bool",
                        name: "",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "bool",
                            },
                        ),
                    },
                ],
                state_mutability: NonPayable,
            },
        ],
        "transferOwnership": [
            Function {
                name: "transferOwnership",
                inputs: [
                    Param {
                        ty: "address",
                        name: "newOwner",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
                outputs: [],
                state_mutability: NonPayable,
            },
        ],
    },
    events: {
        "Approval": [
            Event {
                name: "Approval",
                inputs: [
                    EventParam {
                        ty: "address",
                        name: "owner",
                        indexed: true,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    EventParam {
                        ty: "address",
                        name: "spender",
                        indexed: true,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    EventParam {
                        ty: "uint256",
                        name: "value",
                        indexed: false,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                anonymous: false,
            },
        ],
        "OwnershipTransferred": [
            Event {
                name: "OwnershipTransferred",
                inputs: [
                    EventParam {
                        ty: "address",
                        name: "previousOwner",
                        indexed: true,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    EventParam {
                        ty: "address",
                        name: "newOwner",
                        indexed: true,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
                anonymous: false,
            },
        ],
        "Transfer": [
            Event {
                name: "Transfer",
                inputs: [
                    EventParam {
                        ty: "address",
                        name: "from",
                        indexed: true,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    EventParam {
                        ty: "address",
                        name: "to",
                        indexed: true,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    EventParam {
                        ty: "uint256",
                        name: "value",
                        indexed: false,
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
                anonymous: false,
            },
        ],
    },
    errors: {
        "ERC20ExceededCap": [
            Error {
                name: "ERC20ExceededCap",
                inputs: [
                    Param {
                        ty: "uint256",
                        name: "increasedSupply",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "cap",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InsufficientAllowance": [
            Error {
                name: "ERC20InsufficientAllowance",
                inputs: [
                    Param {
                        ty: "address",
                        name: "spender",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "allowance",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "needed",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InsufficientBalance": [
            Error {
                name: "ERC20InsufficientBalance",
                inputs: [
                    Param {
                        ty: "address",
                        name: "sender",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "balance",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                    Param {
                        ty: "uint256",
                        name: "needed",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InvalidApprover": [
            Error {
                name: "ERC20InvalidApprover",
                inputs: [
                    Param {
                        ty: "address",
                        name: "approver",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InvalidCap": [
            Error {
                name: "ERC20InvalidCap",
                inputs: [
                    Param {
                        ty: "uint256",
                        name: "cap",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "uint256",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InvalidReceiver": [
            Error {
                name: "ERC20InvalidReceiver",
                inputs: [
                    Param {
                        ty: "address",
                        name: "receiver",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InvalidSender": [
            Error {
                name: "ERC20InvalidSender",
                inputs: [
                    Param {
                        ty: "address",
                        name: "sender",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
            },
        ],
        "ERC20InvalidSpender": [
            Error {
                name: "ERC20InvalidSpender",
                inputs: [
                    Param {
                        ty: "address",
                        name: "spender",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
            },
        ],
        "OwnableInvalidOwner": [
            Error {
                name: "OwnableInvalidOwner",
                inputs: [
                    Param {
                        ty: "address",
                        name: "owner",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
            },
        ],
        "OwnableUnauthorizedAccount": [
            Error {
                name: "OwnableUnauthorizedAccount",
                inputs: [
                    Param {
                        ty: "address",
                        name: "account",
                        components: [],
                        internal_type: Some(
                            Other {
                                contract: None,
                                ty: "address",
                            },
                        ),
                    },
                ],
            },
        ],
    },
}
